### PR TITLE
Feature: Enable default language subpath

### PR DESCRIPTION
**Pull request by anandaprabawa**
This is my proposal to enable default language subpath.

Currently the default language just showing the root path `/`, and we might want to show the language `/en/` in the address bar

To enable this feature by passing an option into the NextI18Next constructor:
```jsx
new NextI18Next({ localeSubpaths: true, defaultLocaleSubpath: true })
```

If `localeSubpaths` is false, the defaultLocaleSubpath won't work.

### CODE REVIEW

**Comment by isaachinman** [comment]
It's a nice proposal, but do you feel like contributing a PR?

**Comment by anandaprabawa** [comment]
> It's a nice proposal, but do you feel like contributing a PR?

Yes, of course

File: src/utils/lng-path-detector.js [hunk 1]
```diff
@@ -23,9 +23,13 @@ export default (req, res, next) => {
     }
     /*
       If a user has a default language prefix
-      in their URL, strip it.
+      in their URL, strip it. Except it has defaultLocaleSubpath to true
```
Comment (isaachinman) [inline src/utils/lng-path-detector.js]: Let's instead write:

> If a user has a default language prefix in their URL and `defaultLocaleSubpath` set to false, strip it.

Comment (anandaprabawa) [reply src/utils/lng-path-detector.js]: I've updated this and the readme.

**Comment by isaachinman** [comment]
Ah, please ignore that. I was mistaken. Another consideration, though: in the new options paradigm, `localeSubpaths` and `defaultLocaleSubpath` are interdependent. That is to say, the `defaultLocaleSubpath` option requires that the `localeSubpaths` option also be true.

Because of that, I am strongly considering changing the option from a bool to a string, where `localeSubpaths` can either be `non-defaults` or `all`. This would of course be a breaking change, so perhaps it's best to line that up with a v1.0 release.

Thanks again.

**Comment by isaachinman** [comment]
@anandaprabawa If, given this new option, I hit `localhost:3000` with a default language of `en`, the server actually returns a 200 with our Next doc. The redirect only takes place on the client. This is not a production-ready solution - we would need to return a 301 to `localhost:3000/en/`.

There are already 301s in place in some of the source code, but I had not considered this use case. I think we're going to need some new logic.

**Comment by isaachinman** [comment]
Any update here?

**Comment by anandaprabawa** [comment]
> Any update here?

Not yet, can we close this PR?

**Comment by isaachinman** [comment]
Why do you want to close it? We're 90% of the way there.

**Comment by anandaprabawa** [comment]
@isaachinman i've updated the changes to return 301 when hit `localhost:3000`, does it work as you expected?

File: src/utils/lng-path-detector.js [hunk 2]
```diff
@@ -21,11 +23,23 @@ export default (req, res, next) => {
...
       req.i18n.changeLanguage(defaultLanguage)
     }
+
     /*
-      If a user has a default language prefix
-      in their URL, strip it.
+      If a user enable the localeSubpaths and defaultLocaleSubpath,
+      request to the root path will redirect to default language
+     */
+    if (localeSubpaths && defaultLocaleSubpath && req.url === '/') {
```
Comment (isaachinman) [inline src/utils/lng-path-detector.js]: You need to check here if the current language is the default language, no?

**Comment by isaachinman** [comment]
This PR will conflict with #48 - one of you will have to do some conflict resolution.

File: src/utils/lng-path-detector.js (hunk 2 above)
Comment (anandaprabawa) [reply src/utils/lng-path-detector.js]: Isn't that the default language always the root path `/` ? why we need to check it?

Comment (isaachinman) [reply src/utils/lng-path-detector.js]: Yes actually, this should be fine. However, you should not be using `req.url` as that will include query parameters.

**Comment by isaachinman** [comment]
Anyone interested in picking this up? Otherwise it will be closed as it's quite old now.

**Comment by capellini** [comment]
@isaachinman I can pick it up, but if possible, given how much the code base has changed, could we do so after #90 is merged?

**Comment by isaachinman** [comment]
@capellini Sure.

**Comment by lucasfeliciano** [comment]
I can also help it out.
Let me know if you easily can tackle this @capellini, otherwise I'm up to jump into this too.

**Comment by capellini** [comment]
@lucasfeliciano Acutally, if you don't mind taking over this one, I'd like to focus on getting unit tests to 100% coverage and getting more e2e tests up and running.

**Comment by lucasfeliciano** [comment]
You can also assign me to this one @isaachinman.
I would appreciate if you assign me to this issue; otherwise, I lose focus. 🙈

**Comment by isaachinman** [comment]
@lucasfeliciano This one I will gladly assign to you, but I think we're better off closing this PR and opening an issue to track the progress, as this work is very old at this point.

**Comment by lucasfeliciano** [comment]
Alright!

Let me know about the priority on this issue because I'm going on a city trip on Thursday and I'm on Sunday.
I can tackle this when I come back if it is not a problem

**Comment by isaachinman** [comment]
@lucasfeliciano Priority on this is fairly low to be honest. A couple people have requested this, but have disappeared or not responded when I tried to continue the discussion. There are probably other areas where your help would be more useful.

**Comment by dayze** [comment]
We will be very interested in this feature, I can try to look at it this week.

**Comment by chemuto** [comment]
Hi @dayze . Did you have a chance to look into this? I would also be interested in using this feature. Cheers

**Comment by dayze** [comment]
Hey @chemuto, I'll start on it monday.

**Pull request closed without merging**